src/viam/app/data/client.py

```diff
@@ -41,37 +55,149 @@ def __init__(self, channel: Channel, metadata: str):
+        return data
 
     async def binary_data_by_filter(
         self,
-        data_request: Optional[Filter],
-        dest: Optional[str]
+        filter: Optional[Filter] = None,
+        dest: Optional[str] = None,
     ) -> List[bytes]:
-        raise NotImplementedError()
+        """Filter and download binary data
 
-    async def binary_data_by_ids(self, file_ids: Optional[List[str]]) -> List[bytes]:
-        raise NotImplementedError()
+        Args:
+            filter (viam.app.data.Filter): When supplied, the binary data will be
+            filtered based on the provided constraints. Otherwise, all data is returned.
+            dest (str): When supplied, the binary data will be saved to the provided file path
+
+        Returns:
+            List[bytes]: The binary data
+        """
+        filter = filter if filter else Filter()
+        last = ""
+        data = []
+
+        # `DataRequest`s are limited to 100 pieces of data, so we loop through calls until
+        # we are certain we've received everything.
+        while True:
+            data_request = DataRequest(filter=filter, limit=100, last=last)
+            request = BinaryDataByFilterRequest(data_request=data_request, count_only=False)
+            response: BinaryDataByFilterResponse = await self._data_client.BinaryDataByFilter(request, metadata=self._metadata)
+            if not response.data or len(response.data) == 0:
+                break
+            data += list(response.data)
+            last = response.last
+
+        if dest:
+            try:
+                file = open(dest, 'w')
+                file.write(f"{data}")
+            except Exception as e:
+                LOGGER.error(f"Failed to write binary data to file {dest}", exc_info=e)
+
+        return data
+
+    async def binary_data_by_ids(
+        self,
+        binary_ids: List[BinaryID],
+        dest: Optional[str] = None,
+    ) -> List[bytes]:
+        """Filter and download binary data
+
+        Args:
+            binary_ids (List[viam.proto.app.BinaryID]): IDs of the desired data. Must be non-empty.
+            dest (str): When supplied, the binary data will be saved to the provided file path.
+
+        Returns:
+            List[bytes]: The binary data.
+
+        Raises:
+            GRPCError: if no binary_ids are provided.
+        """
+        request = BinaryDataByIDsRequest(binary_ids=binary_ids, include_binary=True)
+        response: BinaryDataByIDsResponse = await self._data_client.BinaryDataByIDs(request, metadata=self._metadata)
+        if dest:
+            try:
+                file = open(dest, 'w')
+                file.write(f"{response.data}")
+            except Exception as e:
+                LOGGER.error(f"Failed to write binary data to file {dest}", exc_info=e)
+        return response.data
 
     async def delete_tabular_data_by_filter(self, filter: Optional[Filter]) -> None:
```

Comment:
Good question! I meant to leave a comment to this effect but forgot to. I opted to not providing a default for the `delete_` methods because I don't want to make it easier for people to delete everything. By not providing a default, users are forced to create a filter. This still allows for a "delete everything" case but hopefully makes users think a bit before doing so.

Reply:
Love it